backends/tofino/bf-p4c/mau/gateway.cpp

```diff
@@ -160,7 +160,7 @@ const IR::Expression *CanonGatewayExpr::postorder(IR::Operation::Relation *e) {
     Pattern::Match<IR::Expression> e1, e2;
     Pattern::Match<IR::Constant> k1, k2;
     int width = safe_width_bits(e->left->type);
-    if (((e1 & k1) == k2).match(e) || ((e1 & k1) != k2).match(e)) {
+    if (((e1 & k1).operator==(k2)).match(e) || ((e1 & k1).operator==(k2)).match(e)) {
```

Comment:
Wow, this makes pattern usage quite ugly...

Reply:
There was some funkyness the way operator== is implemented for the Pattern class: https://github.com/p4lang/p4c/blob/main/ir/pattern.h#L101

Let me see whether there is a better way to resolve this.

Reply:
Or maybe using alternative pattern matcher from https://github.com/p4lang/p4c/pull/5284 :)

Reply:
I had some fix for this somewhere (it might be lost now :-/) I think it was necessary to explicitly add the overloads implied previously by implicit cast. The reason is the implicit cast play badly with the C++20's ability to swap around operands of `==`.

Reply:
Maybe need to add an explicit overload

    template<class T> Pattern operator==(Match<T> &a) { return Pattern(*this) == Pattern(a); }

to get it to not do whatever was causing any problems you were seeing.  But putting explicit `operator==` in the *use* of the pattern defeats the whole purpose (which is to be able to have a code fragment that looks like P4 code and say "does the IR match this code fragment?")

Reply:
I found I could make these work with GCC-11 by swapping the arguments:

    if ((k2 == (e1 & k1)).match(e) || (k2 != (e1 & k1)).match(e)) {

that works because Pattern::Match<IR::Equ> (and Neq) are commutative and will match swapped operands.